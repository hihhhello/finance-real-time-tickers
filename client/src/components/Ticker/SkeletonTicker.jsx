import React from "react";
import classNames from "classnames";
import styles from "./ticker.module.scss";

export const SkeletonTicker = ({ ticker }) => (
  <div className={classNames(styles.wrapper, styles.wrapperSkeleton)}>
    <div
      className={classNames(styles.tickWrapper, styles.tickWrapperSkeleton)}
    ></div>
    <div className={classNames(styles.textWrapper, styles.textWrapperSkeleton)}>
      <span
        className={classNames(styles.name, {
          [styles.nameSkeleton]: !ticker && true,
        })}
      >
        {ticker}
      </span>
      <span className={classNames(styles.price, styles.priceSkeleton)}></span>
    </div>
    <div
      className={classNames(styles.changeWrapper, styles.changeWrapperSkeleton)}
    >
      <span
        className={classNames(
          styles.changePercent,
          styles.changePercentSkeleton
        )}
      ></span>
      <span className={classNames(styles.change, styles.changeSkeleton)}></span>
    </div>
  </div>
);
