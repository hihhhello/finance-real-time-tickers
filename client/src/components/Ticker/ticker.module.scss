@import "../../scss/var";
@import "../../scss/mixin";

@include skeleton-keyframes;
@include scroll-animation-settings;
@include fades-keyframes;

.wrapper {
  position: relative;
  color: $green-color;
  display: flex;
  align-items: center;
  font-size: 0.071em; // ~12px (depends on fz of grid container)
  animation: fade-in 500ms alternate ease-out;
  gap: 0.65em;
  padding: 0.5em;
  background-color: #fff;
  border: 1px solid $border-color;
  border-radius: 0.5em;
  cursor: pointer;
  &:hover {
    background-color: $hover-color;
  }
  &Skeleton {
    pointer-events: none;
  }
  &Loss {
    color: $red-color;
  }
  &Paused {
    filter: grayscale(100%);
  }
}

.tickWrapper {
  display: grid;
  place-items: center;
  min-height: 2.7em;
  min-width: 2.7em;
  background-color: $light-green-color;
  border-radius: 0.35em;
  &Loss {
    background-color: $light-red-color;
  }
  &Skeleton {
    @include skeleton-settings;
  }
}

.tick {
  max-height: 1.4em;
  max-width: 1.4em;
}

.textWrapper {
  text-align: left;
  display: flex;
  flex-direction: column;
  min-width: 5.3em;
  max-width: 10em;
  &Skeleton {
    min-height: 100%;
    justify-content: space-around;
  }
}

.priceWrapper {
  max-height: 1em;
  overflow: hidden;
}

.changePriceWrapper,
.changePercentWrapper {
  overflow: hidden;
}

.price,
.name {
  overflow: hidden;
  color: $primary-font-color;
  &Skeleton {
    height: 0.8em;
    width: 100%;
    border-radius: 100em;
    @include skeleton-settings;
  }
}

.name {
  text-transform: uppercase;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-weight: 700;
  &Skeleton {
    width: 50%;
  }
}

.changeWrapper {
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: right;
  flex: 2;
  &Skeleton {
    height: 100%;
    justify-content: space-evenly;
  }
}

.changePercent,
.change {
  overflow: hidden;
  &Skeleton {
    height: 0.5em;
    width: 60%;
    max-width: 2.5em;
    border-radius: 100em;
    align-self: flex-end;
    @include skeleton-settings;
    &::after {
      display: none;
    }
  }
}

.changePercent {
  font-weight: 700;
  padding-right: em;
  &::after {
    content: "%";
    margin-left: 0.1em;
  }
}
