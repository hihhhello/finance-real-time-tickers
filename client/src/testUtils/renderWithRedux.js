import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { render } from "@testing-library/react";
import reducers from "../store/slices";

export const renderWithRedux = (
  component,
  {
    initialState,
    store = createStore(combineReducers(reducers), initialState),
  } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};
