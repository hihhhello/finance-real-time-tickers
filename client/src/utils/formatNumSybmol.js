export function formatNumSybmol(num) {
  return +num === 0 || num < 0
    ? num
    : `+${Intl.NumberFormat("ru-RU").format(num)}`;
}
