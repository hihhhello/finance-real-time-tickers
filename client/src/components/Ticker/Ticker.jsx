import React, { useRef } from "react";
import classNames from "classnames";
import { SwitchTransition, CSSTransition } from "react-transition-group";
import { useSelector } from "react-redux";

import { formatNumSybmol } from "../../utils/formatNumSybmol";

import { SkeletonTicker } from "./SkeletonTicker";
import { ReactComponent as ArrowUp } from "../../icons/arrow-up-solid.svg";
import { ReactComponent as ArrowDown } from "../../icons/arrow-down-solid.svg";

import styles from "./ticker.module.scss";

const ScrollAnimation = ({ children, keyValue, elemRef }) => {
  return (
    <SwitchTransition mode={"out-in"}>
      <CSSTransition
        nodeRef={elemRef}
        key={keyValue}
        timeout={200}
        classNames={{
          enterActive: styles.scrollEnterActive,
          exitActive: styles.scrollExitActive,
        }}
      >
        {children}
      </CSSTransition>
    </SwitchTransition>
  );
};

export const Ticker = ({
  ticker,
  // exchange,
  price,
  change,
  changePercent,
  onClickHandler,
  skeleton = false,
}) => {
  const { isWatching } = useSelector((state) => state.userSettings);
  const isLoss = +change && +changePercent < 0;
  const priceRef = useRef();
  const changeRef = useRef();
  const changePercentRef = useRef();
  return (
    <>
      {skeleton ? (
        <SkeletonTicker ticker={ticker} />
      ) : (
        <div
          className={classNames(styles.wrapper, {
            [styles.wrapperLoss]: isLoss,
            [styles.wrapperPaused]: !isWatching,
          })}
          onClick={() => onClickHandler(ticker)}
        >
          <div
            className={classNames(styles.tickWrapper, {
              [styles.tickWrapperLoss]: isLoss,
            })}
          >
            {isLoss ? (
              <ArrowDown className={styles.tick} />
            ) : (
              <ArrowUp className={styles.tick} />
            )}
          </div>
          <div className={styles.textWrapper}>
            <div className={styles.name}>{ticker}</div>
            <div className={styles.priceWrapper}>
              <ScrollAnimation keyValue={price} elemRef={priceRef}>
                <div ref={priceRef} className={styles.price}>
                  {Intl.NumberFormat("ru-RU").format(price)}
                </div>
              </ScrollAnimation>
            </div>
          </div>
          <div className={styles.changeWrapper}>
            <div className={styles.changePercentWrapper}>
              <ScrollAnimation
                elemRef={changePercentRef}
                keyValue={changePercent}
              >
                <div
                  ref={changePercentRef}
                  className={styles.changePercent}
                >{`${formatNumSybmol(changePercent)}`}</div>
              </ScrollAnimation>
            </div>
            <div className={styles.changePriceWrapper}>
              <ScrollAnimation keyValue={change} elemRef={changeRef}>
                <div ref={changeRef} className={styles.change}>
                  {formatNumSybmol(change)}
                </div>
              </ScrollAnimation>
            </div>
          </div>
        </div>
      )}
    </>
  );
};
