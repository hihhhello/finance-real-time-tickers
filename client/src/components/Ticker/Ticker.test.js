import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Ticker from ".";
import { renderWithRedux } from "../../testUtils/renderWithRedux";
import { formatNumSybmol } from "../../utils/formatNumSybmol";

const tickerData = {
  ticker: "TSL",
  price: 234.7,
  change: 48.9,
  changePercent: 0.34,
};

const initialState = {
  userSettings: {
    isWatching: true,
  },
};

const onClick = jest.fn();

describe("Ticker", () => {
  it("Skeleton snapshot", () => {
    const { asFragment, container } = renderWithRedux(<Ticker skeleton />);
    expect(container.firstChild).toHaveClass("wrapperSkeleton");
    expect(asFragment(<Ticker skeleton />)).toMatchSnapshot();
  });

  it("empty snapshot", () => {
    const { asFragment } = renderWithRedux(<Ticker />);

    expect(asFragment(<Ticker />)).toMatchSnapshot();
  });

  it("renders", () => {
    const { getByText, asFragment } = renderWithRedux(
      <Ticker {...tickerData} />,
      {
        initialState,
      }
    );

    expect(getByText(tickerData.ticker)).toBeInTheDocument();
    expect(
      getByText(Intl.NumberFormat("ru-RU").format(tickerData.price))
    ).toBeInTheDocument();
    expect(getByText(formatNumSybmol(tickerData.change))).toBeInTheDocument();
    expect(
      getByText(formatNumSybmol(tickerData.changePercent))
    ).toBeInTheDocument();
    expect(asFragment(<Ticker />)).toMatchSnapshot();
  });

  it("paused on toggle click", () => {
    const { container, asFragment } = renderWithRedux(
      <Ticker {...tickerData} />,
      {
        initialState: {
          userSettings: {
            isWatching: false,
          },
        },
      }
    );
    expect(asFragment(<Ticker />)).toMatchSnapshot();
    expect(container.firstChild).toHaveClass("wrapperPaused");
  });

  it("onClick works", () => {
    const { container } = renderWithRedux(
      <Ticker {...tickerData} onClickHandler={onClick} />,
      {
        initialState,
      }
    );
    userEvent.click(container.firstChild);
    expect(onClick).toHaveBeenCalled();
  });

  it("changed styles for negative numbers", () => {
    const negativeTickerData = {
      ...tickerData,
      price: -234.7,
      change: -48.9,
      changePercent: -0.34,
    };
    const { container } = renderWithRedux(<Ticker {...negativeTickerData} />, {
      initialState,
    });
    expect(container.firstChild).toHaveClass("wrapperLoss");
  });
});
