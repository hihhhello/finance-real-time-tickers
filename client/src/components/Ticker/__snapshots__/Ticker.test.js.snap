// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticker Skeleton snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper wrapperSkeleton"
  >
    <div
      class="tickWrapper tickWrapperSkeleton"
    />
    <div
      class="textWrapper textWrapperSkeleton"
    >
      <span
        class="name nameSkeleton"
      />
      <span
        class="price priceSkeleton"
      />
    </div>
    <div
      class="changeWrapper changeWrapperSkeleton"
    >
      <span
        class="changePercent changePercentSkeleton"
      />
      <span
        class="change changeSkeleton"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Ticker empty snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper wrapperPaused"
  >
    <div
      class="tickWrapper"
    >
      <svg
        class="tick"
      >
        arrow-up-solid.svg
      </svg>
    </div>
    <div
      class="textWrapper"
    >
      <div
        class="name"
      />
      <div
        class="priceWrapper"
      >
        <div
          class="price"
        >
          не число
        </div>
      </div>
    </div>
    <div
      class="changeWrapper"
    >
      <div
        class="changePercentWrapper"
      >
        <div
          class="changePercent"
        >
          +не число
        </div>
      </div>
      <div
        class="changePriceWrapper"
      >
        <div
          class="change"
        >
          +не число
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Ticker paused on toggle click 1`] = `
<DocumentFragment>
  <div
    class="wrapper wrapperPaused"
  >
    <div
      class="tickWrapper"
    >
      <svg
        class="tick"
      >
        arrow-up-solid.svg
      </svg>
    </div>
    <div
      class="textWrapper"
    >
      <div
        class="name"
      >
        TSL
      </div>
      <div
        class="priceWrapper"
      >
        <div
          class="price"
        >
          234,7
        </div>
      </div>
    </div>
    <div
      class="changeWrapper"
    >
      <div
        class="changePercentWrapper"
      >
        <div
          class="changePercent"
        >
          +0,34
        </div>
      </div>
      <div
        class="changePriceWrapper"
      >
        <div
          class="change"
        >
          +48,9
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Ticker renders 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="tickWrapper"
    >
      <svg
        class="tick"
      >
        arrow-up-solid.svg
      </svg>
    </div>
    <div
      class="textWrapper"
    >
      <div
        class="name"
      >
        TSL
      </div>
      <div
        class="priceWrapper"
      >
        <div
          class="price"
        >
          234,7
        </div>
      </div>
    </div>
    <div
      class="changeWrapper"
    >
      <div
        class="changePercentWrapper"
      >
        <div
          class="changePercent"
        >
          +0,34
        </div>
      </div>
      <div
        class="changePriceWrapper"
      >
        <div
          class="change"
        >
          +48,9
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
